near/omni-token/src/lib.rs

```diff
@@ -0,0 +1,275 @@
+use near_contract_standards::fungible_token::metadata::{
+    FungibleTokenMetadata, FungibleTokenMetadataProvider, FT_METADATA_SPEC,
+};
+use near_contract_standards::fungible_token::{
+    FungibleToken, FungibleTokenCore, FungibleTokenResolver,
+};
+use near_contract_standards::storage_management::{
+    StorageBalance, StorageBalanceBounds, StorageManagement,
+};
+use near_sdk::collections::LazyOption;
+use near_sdk::json_types::{Base64VecU8, U128};
+use near_sdk::serde_json::json;
+use near_sdk::{
+    env, ext_contract, near, require, AccountId, Gas, NearToken, PanicOnDefault, Promise,
+    PromiseOrValue, PublicKey, StorageUsage,
+};
+use omni_types::{BasicMetadata, OmniAddress};
+const OUTER_UPGRADE_GAS: Gas = Gas::from_tgas(15);
+const NO_DEPOSIT: NearToken = NearToken::from_yoctonear(0);
+const CURRENT_STATE_VERSION: u32 = 1;
+
+#[near(contract_state)]
+#[derive(PanicOnDefault)]
+pub struct OmniToken {
+    controller: AccountId,
+    token: FungibleToken,
+    metadata: LazyOption<FungibleTokenMetadata>,
+}
+
+#[ext_contract(ext_omni_factory)]
+pub trait ExtOmniTokenFactory {
+    fn init_transfer(
+        &self,
+        sender: AccountId,
+        amount: U128,
+        recipient: OmniAddress,
+        fee: U128,
+        native_fee: U128,
+    ) -> Promise;
+}
+
+#[near]
+impl OmniToken {
+    #[init]
+    pub fn new(controller: AccountId, metadta: BasicMetadata) -> Self {
+        let current_account_id = env::current_account_id();
+        let deployer_account = current_account_id
+            .get_parent_account_id()
+            .unwrap_or_else(|| env::panic_str("ERR_INVALID_PARENT_ACCOUNT"));
+
+        require!(
+            env::predecessor_account_id().as_str() == deployer_account,
+            "Only the deployer account can init this contract"
+        );
+
+        Self {
+            controller,
+            token: FungibleToken::new(b"t".to_vec()),
+            metadata: LazyOption::new(
+                b"m".to_vec(),
+                Some(&FungibleTokenMetadata {
+                    spec: FT_METADATA_SPEC.to_string(),
+                    name: metadta.name,
+                    symbol: metadta.symbol,
+                    icon: None,
+                    reference: None,
+                    reference_hash: None,
+                    decimals: metadta.decimals,
+                }),
+            ),
+        }
+    }
+
+    pub fn set_metadata(
```

Comment:
How is this going to be called?

Reply:
716e257d59cdabc38e532b536eaa886b344a7a3d